Source/WebKit/UIProcess/Extensions/WebExtension.cpp

```diff
@@ -0,0 +1,2031 @@
+#elif PLATFORM(GTK)
+            const auto linuxPlatform = "linux"_s;
+#endif
+            const auto defaultPlatform = "default"_s;
+
+#if PLATFORM(MAC)
+            auto platformShortcut = !suggestedKeyDictionary->getString(macPlatform).isEmpty() ? suggestedKeyDictionary->getString(macPlatform) : suggestedKeyDictionary->getString(iosPlatform);
+#elif PLATFORM(IOS_FAMILY)
+            auto platformShortcut = !suggestedKeyDictionary->getString(iosPlatform).isEmpty() ? suggestedKeyDictionary->getString(iosPlatform) : suggestedKeyDictionary->getString(macPlatform);
+#elif PLATFORM(GTK)  
+            auto platformShortcut = suggestedKeyDictionary->getString(linuxPlatform);
+#else
+            auto platformShortcut = suggestedKeyDictionary->getString(defaultPlatform);
+#endif
+            if (platformShortcut.isEmpty())
+                platformShortcut = suggestedKeyDictionary->getString(defaultPlatform);
+            
+            if (!parseCommandShortcut(platformShortcut, commandData.modifierFlags, commandData.activationKey)) {
+                error = WEB_UI_STRING("Invalid `suggested_key` in the `commands` manifest entry.", "WKWebExtensionErrorInvalidManifestEntry description for invalid command shortcut");
+                continue;
+            }
+
+            if (!commandData.activationKey.isEmpty() && ++commandsWithShortcuts > maximumNumberOfShortcutCommands) {
+                error = WEB_UI_STRING("Too many shortcuts specified for `commands`, only 4 shortcuts are allowed.", "WKWebExtensionErrorInvalidManifestEntry description for too many command shortcuts");
+                commandData.activationKey = emptyString();
+                commandData.modifierFlags = { };
+            }
+        }
+
+        m_commands.append(WTFMove(commandData));
+    }
+
+    if (!hasActionCommand) {
+        String commandIdentifier;
+        if (hasAction())
+            commandIdentifier = "_execute_action"_s;
+        else if (hasBrowserAction())
+            commandIdentifier = "_execute_browser_action"_s;
+        else if (hasPageAction())
+            commandIdentifier = "_execute_page_action"_s;
+
+        if (!commandIdentifier.isEmpty())
+            m_commands.append({ commandIdentifier, displayActionLabel(), emptyString(), { } });
+    }
+
+    if (error)
+        recordError(createError(Error::InvalidCommands, error.value()));
+}
+
+std::optional<WebExtension::DeclarativeNetRequestRulesetData> WebExtension::parseDeclarativeNetRequestRulesetDictionary(RefPtr<JSON::Value> rulesetDictionary, RefPtr<API::Error> *error)
+{
+    std::array<String, 3> requiredKeysInRulesetDictionary = {
+        declarativeNetRequestRulesetIDManifestKey,
+        declarativeNetRequestRuleEnabledManifestKey,
+        declarativeNetRequestRulePathManifestKey
+    };
+
+    // FIXME: Validate Dictionary (will be done when WebExtensionUtilities is ported)
+
+    String rulesetID = rulesetDictionary->asObject()->getString(declarativeNetRequestRulesetIDManifestKey);
```

Comment:
`nullptr`

EDIT: I guess your comment above implies it will be validated ahead of time.

Reply:
yes, but the old validator is still in Cocoa and hasn't been ported to work with JSON::Object yet. however, we can certainly do a temporary nullptr check here for now